CheatsManager: Create Action Replay and Gecko code widgets only once

Dentomologist opened the pull request:
Create CheatsManager's ARCodeWidget and GeckoCodeWidget once on startup (in CheatsManager's constructor), and trigger updates to their contents when the current game changes.

Previously, ARCodeWidget and GeckoCodeWidget would be recreated every time a game was started or shutdown. If either of these widgets was the visible tab this would cause the tab to switch to CheatSearchFactoryWidget (since the other tabs temporarily didn't exist), and also could cause a crash if the user initiated a game shutdown and then opened a code edit window before the shutdown finished.

Tilka reviewed (approved):
Sounds/looks good. Untested.

mitaclaw commented:
I can confirm the issue wherein the `CheatsManager` switches to the `CheatSearchFactoryWidget` whenever emulation is started or stopped has been fixed.

The issue wherein a code edit window open during shutdown causes a crash (abort due to `free(): invalid pointer`) has been fixed, but in its place I have found a new issue: attempting to save a cheat code after emulation has stopped will crash (abort due to `double free or corruption (out)`). Pressing the "Cancel" push button or closing the window does not crash.

Saving a cheat code when no software is running doesn't make much sense in the first place—it would probably write to an INI for the default GameID, 00000000—so I think this action should be discouraged in the UI if possible.

Dentomologist commented:
I've pushed an update. The most important change is that when the game changes, the CheatCodeEditor in Cheats Manager's ARCodeWidget and GeckoCodeWidget are rejected and so can't be used to save a code in an invalid context. The widgets in a game's Properties window stay open and continue working correctly.

A few other changes that I made either out of necessity or because they were right there:
* Removed the calls to SetQWidgetWindowDecorations. As far as my testing has shown they aren't actually necessary for the CheatCodeEditor to respond to themes correctly, and in combination with the changes I had to make to enable rejecting the editor they were causing the Cheats Manager window to be far too small on creation.
* The Edit and Remove buttons in ARCodeWidget are now disabled when no code is selected.
* Some minor refactoring and constifying.
* Removed a call to OnSelectionChange that had no effect in context.

mitaclaw commented on Source/Core/DolphinQt/CheatsManager.cpp:
```diff
@@ -172,14 +148,9 @@ void CheatsManager::OnNewSessionCreated(const Cheats::CheatSearchSessionBase& se
-  w->connect(w, &CheatSearchWidget::RequestWatch,
-             [this](QString name, u32 address) { emit RequestWatch(name, address); });
+  connect(w, &CheatSearchWidget::ActionReplayCodeGenerated, m_ar_code, &ARCodeWidget::AddCode);
+  connect(w, &CheatSearchWidget::ShowMemory, this, &CheatsManager::ShowMemory);
+  connect(w, &CheatSearchWidget::RequestWatch, this, &CheatsManager::RequestWatch);
```
I'm not sure if I agree with directly calling signals as if they were slots. If I understand, the `emit` keyword is syntactic sugar and nothing more, but it's still helpful self-documentation.

Dentomologist replied:
The [Qt docs](https://doc.qt.io/qt-6/qobject.html#connect) say "A signal can also be connected to another signal" and give this (abbreviated) example:

```
myButton = new QPushButton(this);
connect(myButton, SIGNAL(clicked()), this, SIGNAL(buttonClicked()));
```

If we want to send a signal manually (i.e. not in a `connect` call) then I agree with using the `emit` keyword for clarity even though it's not necessary, but for `connect` I think wrapping it in a lambda just to use `emit` would make more noise than light.

mitaclaw replied:
Well I can't argue with the official docs. I learn a new thing about Qt every day.

mitaclaw commented:
Please restore the `SetQWidgetWindowDecorations` call before each `QDialog` `exec` call. Without it, the title bar will (I think?) be the system default, which may or may not match the current window style if Dolphin is configured to override it with light or dark style. I was unable to reproduce the too-small dialog window you described, unless I just forgot to reinstate a  `SetQWidgetWindowDecorations` call in a key place?

![With SetQWidgetWindowDecorations](https://github.com/user-attachments/assets/3224333e-e69f-4058-a765-8134b61818d8) ![Without SetQWidgetWindowDecorations](https://github.com/user-attachments/assets/9e79f655-94ec-43b7-915a-422eb713df69)
Here are screenshots in which the system default style is light mode but Dolphin overrides it with dark mode. On the left is with the `SetQWidgetWindowDecorations` call, on the right is without.

Dentomologist commented:
I believe the too-small window was a result of my attempts to avoid having to call `SetQWidgetWindowDecorations` on `m_cheat_code_editor` every time it called `exec` despite the editor being created a single time per AR/GeckoCodeWidget per Dolphin session, but if that's what it takes to make everything display correctly then so be it.

My testing of the removal of `SetQWidgetWindowDecorations` was hampered by the fact that I didn't notice it only affects Windows Light -> Dolphin Dark (while my Windows theme is dark), and also that my Windows version is too old for the function to actually do anything at all (as DWMWINDOWATTRIBUTE::DWMWA_USE_IMMERSIVE_DARK_MODE is unsupported).

Since I can't test it myself, let me know if the theme is being properly applied now.

mitaclaw reviewed (approved):
The title bar issue is fixed, and the two resolved reviews I left were my only concerns.

JMC47 commented:
I was just using cheat search a few days ago and these are some welcome improvements.
